pkg/objstore/caching/caching_bucket.go

```diff
@@ -0,0 +1,290 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package caching
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"regexp"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/thanos-io/thanos/pkg/cache"
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+// Bucket implementation that provides some caching features, using knowledge about how Thanos accesses data.
+type CachingBucket struct {
+	bucket objstore.Bucket
+	cache  cache.Cache
+
+	// Basic unit used to cache chunks.
+	chunkBlockSize int64
+
+	// Minimum number of cached blocks segment to "break" fetching segment.
+	minCachingBlocksSize int64
+
+	// TTLs for various cache items.
+	chunkObjectSizeTTL time.Duration
+	chunkBlockTTL      time.Duration
+
+	cachedBytes  prometheus.Counter
+	fetchedBytes prometheus.Counter
+}
+
+func NewCachingBucket(b objstore.Bucket, c cache.Cache, reg prometheus.Registerer) (*CachingBucket, error) {
+	if b == nil {
+		return nil, errors.New("bucket is nil")
+	}
+	if c == nil {
+		return nil, errors.New("cache is nil")
+	}
+
+	return &CachingBucket{
+		bucket:         b,
+		cache:          c,
+		chunkBlockSize: 16000, // equal to max chunk size
+
+		minCachingBlocksSize: 128 * 1024,
+		chunkObjectSizeTTL:   24 * time.Hour,
+		chunkBlockTTL:        24 * time.Hour,
+
+		cachedBytes: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "thanos_objstore_caching_bucket_cached_chunk_bytes_total",
+			Help: "Total number of chunk blocks used from cache",
+		}),
+		fetchedBytes: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "thanos_objstore_caching_bucket_fetched_chunk_bytes_total",
+			Help: "Total number of chunk blocks fetched from storage",
+		}),
```

Comment:
If we pick the path I've suggested to make this implementation specific to the thanos store (and thus specific cache chunks) we can make these metrics more specific.

Reply:
I've renamed them to `thanos_store_caching_bucket_cached_chunk_bytes_total` and `thanos_store_caching_bucket_fetched_chunk_bytes_total`.